New feature: Favorite cards

t351206 opened the pull request:
Description (this is a follow-up to #305 ):

If you have some favorite loyalty cards it is useful to have them as first items in the listview. Therefore I added the "star-unstar" feature. Cards in the LoyaltyCardViewActivity can be be starred in order to make them a favorite card.

Therefore a new SQLite table column was added where the starring status is saved.

Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I
have the right to submit it under the open source license
indicated in the file; or

(b) The contribution is based upon previous work that, to the best
of my knowledge, is covered under an appropriate open source
license and I have the right under that license to submit that
work with modifications, whether created in whole or in part
by me, under the same open source license (unless I am
permitted to submit under a different license), as indicated
in the file; or

(c) The contribution was provided directly to me by some other
person who certified (a), (b) or (c) and I have not modified
it.

(d) I understand and agree that this project and the contribution
are public and that a record of the contribution (including all
personal information I submit with it, including my sign-off) is
maintained indefinitely and may be redistributed consistent with
this project or the open source license(s) involved.

t351206@gmx.de

TheLastProject commented on app/src/main/java/protect/card_locker/LoyaltyCardViewActivity.java:
```diff
@@ -344,11 +352,42 @@ public boolean onOptionsItemSelected(MenuItem item)
+                    setStarinDB(item, loyaltyCardId, true);
+                }
+                starred = !starred;
+                return true;
```
This whole thing can probably be simplified as:

```
starred = !starred;
setStarInDB(item, loyaltyCardId, starred);
return true;
```

t351206 replied:
Correct! It was just copied from similar case R.id.action_lock_unlock above ....
I changed the code

TheLastProject commented on app/src/test/java/protect/card_locker/LoyaltyCardViewActivityTest.java:
```diff
@@ -488,12 +488,13 @@ public void checkMenu() throws IOException
         assertEquals("Block Rotation", menu.findItem(R.id.action_lock_unlock).getTitle().toString());
         assertEquals("Share", menu.findItem(R.id.action_share).getTitle().toString());
         assertEquals("Edit", menu.findItem(R.id.action_edit).getTitle().toString());
+        assertEquals("Remove from favorites", menu.findItem(R.id.action_star_unstar).getTitle().toString());
```
But starred is 0 in this loyalty card. Shouldn't this be "Add to favorites" then?

t351206 replied:
You are right! This is alraedy wrong in LoyaltyCardViewActivity.java

t351206 commented on app/src/main/java/protect/card_locker/CsvDatabaseImporter.java:
```diff
@@ -142,6 +142,10 @@ private void importLoyaltyCard(SQLiteDatabase database, DBHelper helper, CSVReco
         }
 
-        helper.insertLoyaltyCard(database, id, store, note, cardId, barcodeType, headerColor, headerTextColor);
+        //Starred status will not be imported
```
Good point. In case you export/import the whole database it probably makes sense to import the starred status.
Just if you share only one card with someone else it probably makes no sense (this was my first thought). I can add the code for import of starred status.

TheLastProject replied:
If you just share a single card using the build-in share feature, it doesn't share all fields anyway, just its own whitelist. So for CSV I would really just always import and save the starred status: https://github.com/brarcher/loyalty-card-locker/blob/master/app/src/main/java/protect/card_locker/ImportURIHelper.java

t351206 replied:
I think it's now working correctly - CSV contains the starStatus and single card starStatus is not imported

t351206 commented on app/src/main/java/protect/card_locker/DBHelper.java:
```diff
@@ -116,6 +127,17 @@ public boolean updateLoyaltyCard(final int id, final String store, final String
         return (rowsUpdated == 1);
     }
 
+    public boolean updateLoyaltyCard(final int id, final int starred)
```
Is it really better to merge it with the other updateLoyaltyCard function? Renaming makes sense to me.

TheLastProject replied:
I looked at more of your code and I understand why you want to keep it separate (performance) so a rename would maybe be more sensical, yeah.

t351206 commented on app/src/main/java/protect/card_locker/ImportURIHelper.java:
```diff
@@ -38,11 +41,15 @@ public LoyaltyCard parse(Uri uri) throws InvalidObjectException {
             String cardId = uri.getQueryParameter(CARD_ID);
             String barcodeType = uri.getQueryParameter(BARCODE_TYPE);
+            if (uri.getBooleanQueryParameter(STARRED, false) == false) starred = 0;
+                else starred = 1;
```
thanks, agreed!
In the end when starred is going to SQLite I need an int since SQLite does not have a separate Boolean storage class. Option could be to only use int instead of boolean?

t351206 replied:
not relevant any longer since starStatus is not imported

t351206 commented on app/src/main/java/protect/card_locker/ImportURIHelper.java:
```diff
@@ -38,11 +41,15 @@ public LoyaltyCard parse(Uri uri) throws InvalidObjectException {
             // These values are allowed to be null
             Integer headerColor = null;
             Integer headerTextColor = null;
+            Integer starred;
```
agree, will be corrected

t351206 replied:
deleted, since starStatus is not imported for single cards

t351206 commented on app/src/androidTest/java/protect/card_locker/ApplicationTest.java:
```diff
@@ -1,15 +1,18 @@
     }
-}
\ No newline at end of file
+}*/
```
Yes, this already existing test is not working on my machine (deprecated in API level 24) and I do not know how to solve this issue on my machine. 
It was not planned to push this to github ...  I will revert this

t351206 commented:
> I also want to state that I'm missing tests for this starring functionality.
Which test do you think would make sense?

TheLastProject commented:
I would like to see the following tests:
1. UI shows correct buttons and buttons cause correct state change (once for starring and once for unstarring)
2. Import CSV file with no star information
3. Import with star information
4. Import with invalid star information
5. Main screen showing starred cards on top correctly
6. Import from URI
7. Import from URI with bad data

I think that should cover things

t351206 commented:
@TheLastProject : Thanks for the test proposals!

ad 1) I think I need some help here. 
ad 2) see importWithNoStarredField
ad 3) see multipleCardsExportImportSomeStarred and also implicitly included in other tests
ad 4) see importWithInvalidStarField
ad 5) see addFourLoyaltyCardsTwoStarred
ad 6) no starStatus will be exported for single cards, starStatus will be automatically set to 0 for import; covered with existing test
ad 7) I think this is an already existing issue. There is nothing implemented to avoid import of bad data (store, note, cardId, barcode can be null). I fixed it with an additional check if one of the four values is null. Furthermore I added two tests.

t351206 commented:
How can I test clicking on the star?

shadowOf(activity).clickMenuItem(R.id.action_star_unstar) is producing an exception:
`java.lang.NullPointerException
	at org.robolectric.fakes.RoboMenuItem.setIcon(RoboMenuItem.java:107)
	at protect.card_locker.LoyaltyCardViewActivity.setStarInDB(LoyaltyCardViewActivity.java:370)
	at protect.card_locker.LoyaltyCardViewActivity.onOptionsItemSelected(LoyaltyCardViewActivity.java:359)
	at android.app.Activity.onMenuItemSelected(Activity.java:2914)
	at androidx.fragment.app.FragmentActivity.onMenuItemSelected(FragmentActivity.java:384)
	at androidx.appcompat.app.AppCompatActivity.onMenuItemSelected(AppCompatActivity.java:219)
	at org.robolectric.shadows.ShadowActivity.clickMenuItem(ShadowActivity.java:502)
	at protect.card_locker.LoyaltyCardViewActivityTest.checkPushStarIcon(LoyaltyCardViewActivityTest.java:718)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:50)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:47)
	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
	at org.robolectric.RobolectricTestRunner$HelperTestRunner$1.evaluate(RobolectricTestRunner.java:600)
	at org.junit.internal.runners.statements.RunBefores.evaluate(RunBefores.java:26)
	at org.robolectric.internal.SandboxTestRunner$2.evaluate(SandboxTestRunner.java:260)
	at org.robolectric.internal.SandboxTestRunner.runChild(SandboxTestRunner.java:130)
	at org.robolectric.internal.SandboxTestRunner.runChild(SandboxTestRunner.java:42)
	at org.junit.runners.ParentRunner$3.run(ParentRunner.java:290)
	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:71)
	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:288)
	at org.junit.runners.ParentRunner.access$000(ParentRunner.java:58)
	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:268)
	at org.robolectric.internal.SandboxTestRunner$1.evaluate(SandboxTestRunner.java:84)
	at org.junit.runners.ParentRunner.run(ParentRunner.java:363)
	at org.junit.runner.JUnitCore.run(JUnitCore.java:137)
	at com.intellij.junit4.JUnit4IdeaTestRunner.startRunnerWithArgs(JUnit4IdeaTestRunner.java:68)
	at com.intellij.rt.junit.IdeaTestRunner$Repeater.startRunnerWithArgs(IdeaTestRunner.java:33)
	at com.intellij.rt.junit.JUnitStarter.prepareStreamsAndStart(JUnitStarter.java:230)
	at com.intellij.rt.junit.JUnitStarter.main(JUnitStarter.java:58)`

t351206 commented:
@TheLastProject: I think I solved the testing issue! Test "checkPushStarIcon" is working as expected!

TheLastProject commented:
Sorry I didn't get back to you in your last reply, things have been a bit chaotic here and I forgot. Let me build your branch and play around with it a bit on my phone :)

TheLastProject commented on app/src/main/java/protect/card_locker/CsvDatabaseImporter.java:
```diff
@@ -142,6 +142,8 @@ private void importLoyaltyCard(SQLiteDatabase database, DBHelper helper, CSVReco
         }
 
-        helper.insertLoyaltyCard(database, id, store, note, cardId, barcodeType, headerColor, headerTextColor);
+        int starStatus = extractInt(DBHelper.LoyaltyCardDbIds.STAR_STATUS, record, false);
```
```
2020-07-30 22:00:27.881 29629-29629/protect.card_locker E/LoyaltyCardLocker: Starting file export with: content://com.android.externalstorage.documents/document/primary%3ALoyaltyCardKeychain.csv%20(1)
2020-07-30 22:00:27.960 29629-29718/protect.card_locker E/LoyaltyCardLocker: Failed to input data
    protect.card_locker.FormatException: Field not used but expected: starstatus
        at protect.card_locker.CsvDatabaseImporter.extractInt(CsvDatabaseImporter.java:91)
        at protect.card_locker.CsvDatabaseImporter.importLoyaltyCard(CsvDatabaseImporter.java:145)
        at protect.card_locker.CsvDatabaseImporter.importData(CsvDatabaseImporter.java:32)
        at protect.card_locker.MultiFormatImporter.importData(MultiFormatImporter.java:38)
        at protect.card_locker.ImportExportTask.performImport(ImportExportTask.java:64)
        at protect.card_locker.ImportExportTask.doInBackground(ImportExportTask.java:121)
        at protect.card_locker.ImportExportTask.doInBackground(ImportExportTask.java:16)
        at android.os.AsyncTask$2.call(AsyncTask.java:333)
        at java.util.concurrent.FutureTask.run(FutureTask.java:266)
        at android.os.AsyncTask$SerialExecutor$1.run(AsyncTask.java:245)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1167)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:641)
        at java.lang.Thread.run(Thread.java:764)
```

Should probably fix this, because it sucks if you make a backup but you can't import it in a newer version. Just set starStatus to 0 if the value isn't found in the backup.

TheLastProject replied:
I would like to suggest solving it this way:

```
diff --git a/app/src/main/java/protect/card_locker/CsvDatabaseImporter.java b/app/src/main/java/protect/card_locker/CsvDatabaseImporter.java
index 5d3f5de..96e8171 100644
--- a/app/src/main/java/protect/card_locker/CsvDatabaseImporter.java
+++ b/app/src/main/java/protect/card_locker/CsvDatabaseImporter.java
@@ -142,8 +142,14 @@ public class CsvDatabaseImporter implements DatabaseImporter
             headerTextColor = extractInt(DBHelper.LoyaltyCardDbIds.HEADER_TEXT_COLOR, record, true);
         }
 
-        int starStatus = extractInt(DBHelper.LoyaltyCardDbIds.STAR_STATUS, record, false);
-        if (starStatus>1 || starStatus<0) throw new FormatException("Wrong Star (Favorite) information listed, but is required");
+        int starStatus = 0;
+        try {
+            starStatus = extractInt(DBHelper.LoyaltyCardDbIds.STAR_STATUS, record, true);
+        } catch (FormatException _) {
+            // This field did not exist in versions 0.28 and before
+            // We catch this exception so we can still import old backups
+        }
+        if (starStatus != 1) starStatus = 0;
         helper.insertLoyaltyCard(database, id, store, note, cardId, barcodeType, headerColor, headerTextColor, starStatus);
     }
 }
```

That way, old backups still work and any invalid value gets parsed as 0.

t351206 replied:
Thanks for the code suggestion! I think this is a very robust and user friendly import now.
In order to avoid NullPointerException a minor change is needed: 
extractInt(DBHelper.LoyaltyCardDbIds.STAR_STATUS, record, **false**);

TheLastProject commented:
Have a few more comments but the code seems to work quite well and I do like it a lot :)

TheLastProject commented:
Hi @t351206,

After talking with @brarcher I ended up forking Loyalty Card Locker. I would love to include your feature in my fork (which I will soon publish after I have a new icon). Could you please make this pull request on https://github.com/TheLastProject/Catima?
